File: src/main/java/io/github/qingmo/json/JSONArray.java
/*
 * MIT License
 *
 * Copyright (c) 2021 qingmo(eagleqingluo@gmail.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package io.github.qingmo.json;

import io.github.qingmo.json.internal.TypeUtils;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;
import java.util.ListIterator;
import java.util.Spliterator;
import java.util.function.UnaryOperator;

public final class JSONArray implements List<Object> {
    private List<Object> list;

    public JSONArray(List<Object> list) {
        this.list = new ArrayList<>();
        this.list.addAll(list);
    }

    public JSONArray(int size) {
        this.list = new ArrayList<>(size);
    }

    public JSONArray() {
        this.list = new ArrayList<>(16);
    }

    public final JSONObject getJSONObject(int index) {
        Object value = this.list.get(index);
        if (value == null) {
            return null;
        }
        if (value instanceof JSONObject) {
            return (JSONObject) value;
        }
        return JSON.parseObject(JSON.toJSONString(value));
    }

    public final JSONArray getJSONArray(int index) {
        Object value = this.list.get(index);
        if (value instanceof JSONArray) {
            return (JSONArray) value;
        }
        if (value instanceof List) {
            return new JSONArray((List) value);
        }
        return null;
    }

    public final Boolean getBoolean(int index) {
        Object value = this.list.get(index);
        return TypeUtils.castToBoolean(value);
    }

    public final boolean getBooleanValue(int index) {
        Object value = this.list.get(index);
        try {
            return TypeUtils.castToBoolean(value);
        } catch (Exception e) {
            return false;
        }
    }

    public final <T> List<T> toJavaList(Class<T> clazz) {
        return JSON.parseArray(this.toJSONString(), clazz);
    }

    public final Byte getByte(int index) {
        Object value = this.list.get(index);
        return TypeUtils.castToByte(value);
    }

    public final byte getByteValue(int index) {
        Object value = this.list.get(index);
        try {
            return TypeUtils.castToByte(value);
        } catch (Exception e) {
            return 0;
        }
    }

    public final Short getShort(int index) {
        Object value = this.list.get(index);
        return TypeUtils.castToShort(value);
    }

    public final short getShortValue(int index) {
        Object value = this.list.get(index);
        try {
            return TypeUtils.castToShort(value);
        } catch (Exception ignore) {
            return 0;
        }
    }

    public final Integer getInteger(int index) {
        Object value = this.list.get(index);
        return TypeUtils.castToInt(value);
    }

    public final int getIntValue(int index) {
        Object value = this.list.get(index);
        try {
            return TypeUtils.castToInt(value);
        } catch (Exception ignore) {
            return 0;
        }
    }

    public final Long getLong(int index) {
        Object value = this.list.get(index);
        return TypeUtils.castToLong(value);
    }

    public final long getLongValue(int index) {
        Object value = this.list.get(index);
        try {
            return TypeUtils.castToLong(value);
        } catch (Exception ignore) {
            return 0L;
        }
    }

    public final Float getFloat(int index) {
        Object value = this.list.get(index);
        return TypeUtils.castToFloat(value);
    }

    public final float getFloatValue(int index) {
        Object value = this.list.get(index);
        try {
            return TypeUtils.castToFloat(value);
        } catch (Exception ignore) {
            return 0.0F;
        }
    }

    public final Double getDouble(int index) {
        Object value = this.list.get(index);
        return TypeUtils.castToDouble(value);
    }

    public final double getDoubleValue(int index) {
        Object value = this.list.get(index);
        try {
            return TypeUtils.castToDouble(value);
        } catch (Exception ignore) {
            return 0.0D;
        }
    }

    public final BigDecimal getBigDecimal(int index) {
        Object value = this.list.get(index);
        return TypeUtils.castToBigDecimal(value);
    }

    public final BigInteger getBigInteger(int index) {
        Object value = this.list.get(index);
        return TypeUtils.castToBigInteger(value);
    }

    public final String getString(int index) {
        Object value = this.list.get(index);
        return TypeUtils.castToString(value);
    }

    public final <T> T getObject(int index, Class<T> clazz) {
        Object value = this.list.get(index);
        if (value == null) {
            return null;
        }
        return JSON.parseObject(JSON.toJSONString(value), clazz);
    }

    public final String toJSONString() {
        return JSON.toJSONString(this.list);
    }

    @Override
    public String toString() {
        return this.toJSONString();
    }

    public int getSize() {
        return this.list.size();
    }

    @Override
    public final int size() {
        return this.getSize();
    }

    @Override
    public boolean isEmpty() {
        return this.list.isEmpty();
    }

    @Override
    public boolean contains(Object element) {
        return this.list.contains(element);
    }

    @Override
    public Iterator<Object> iterator() {
        return this.list.iterator();
    }

    @Override
    public final Object[] toArray() {
        return this.list.toArray();
    }

    @Override
    public boolean add(Object element) {
        return this.list.add(element);
    }

    @Override
    public Object remove(int index) {
        return this.list.remove(index);
    }

    @Override
    public boolean remove(Object element) {
        return this.list.remove(element);
    }

    @Override
    public boolean containsAll(Collection elements) {
        return this.list.containsAll(elements);
    }

    @Override
    public boolean addAll(Collection elements) {
        return this.list.addAll(elements);
    }

    @Override
    public boolean addAll(int index, Collection elements) {
        return this.list.addAll(index, elements);
    }

    @Override
    public boolean removeAll(Collection elements) {
        return this.list.removeAll(elements);
    }

    @Override
    public boolean retainAll(Collection elements) {
        return this.list.retainAll(elements);
    }

    @Override
    public void clear() {
        this.list.clear();
    }

    @Override
    public boolean equals(Object other) {
        return this.list.equals(other);
    }

    @Override
    public int hashCode() {
        return this.list.hashCode();
    }

    @Override
    public Object get(int index) {
        return this.list.get(index);
    }

    @Override
    public Object set(int index, Object element) {
        return this.list.set(index, element);
    }

    @Override
    public void add(int index, Object element) {
        this.list.add(index, element);
    }

    @Override
    public int indexOf(Object element) {
        return this.list.indexOf(element);
    }

    @Override
    public int lastIndexOf(Object element) {
        return this.list.lastIndexOf(element);
    }

    @Override
    public ListIterator<Object> listIterator() {
        return this.list.listIterator();
    }

    @Override
    public ListIterator<Object> listIterator(int index) {
        return this.list.listIterator(index);
    }

    @Override
    public List<Object> subList(int fromIndex, int toIndex) {
        return this.list.subList(fromIndex, toIndex);
    }

    @Override
    public void replaceAll(UnaryOperator operator) {
        this.list.replaceAll(operator);
    }

    @Override
    public void sort(Comparator c) {
        this.list.sort(c);
    }

    @Override
    public Spliterator<Object> spliterator() {
        return this.list.spliterator();
    }

    @Override
    public <T> T[] toArray(T[] a) {
        return this.list.toArray(a);
    }
}

File: src/main/kotlin/io/github/qingmo/json/JSONArray.kt
/*
 * MIT License
 *
 * Copyright (c) 2021 qingmo(eagleqingluo@gmail.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package io.github.qingmo.json

import io.github.qingmo.json.JSON.parseArray
import io.github.qingmo.json.internal.TypeUtils
import java.math.BigDecimal
import java.math.BigInteger
import java.util.ArrayList
import java.util.Comparator
import java.util.Spliterator
import java.util.function.UnaryOperator

class JSONArray : MutableList<Any?> {
    private var list: MutableList<Any?>

    constructor(list: MutableList<Any?>) {
        this.list = list
    }

    constructor(size: Int) {
        list = ArrayList(size)
    }

    constructor() {
        list = ArrayList(16)
    }

    fun getJSONObject(index: Int): JSONObject? {
        val value = list.get(index) ?: return null
        if (value is JSONObject) {
            return value
        }
        return JSON.parseObject(JSON.toJSONString(value))
    }

    fun getJSONArray(index: Int): JSONArray? {
        val value = list[index]
        if (value is JSONArray) {
            return value
        }
        if (value !is List<*>) {
            return null
        }
        val params = value.toMutableList()
        return JSONArray(params)
    }

    fun getBoolean(index: Int): Boolean? {
        val value = get(index)
        return TypeUtils.castToBoolean(value)
    }

    fun getBooleanValue(index: Int): Boolean {
        val value = get(index)
        try {
            return TypeUtils.castToBoolean(value)!!
        } catch (ignore: Exception) {
            return false
        }
    }

    fun <T> toJavaList(clazz: Class<T>?): List<T> {
        return parseArray(toJSONString(), clazz)
    }

    fun getByte(index: Int): Byte? {
        val value = get(index)
        return TypeUtils.castToByte(value)
    }

    fun getByteValue(index: Int): Byte {
        val value = get(index)
        try {
            return TypeUtils.castToByte(value)!!
        } catch (ignore: Exception) {
            return 0
        }
    }

    fun getShort(index: Int): Short? {
        val value = get(index)
        return TypeUtils.castToShort(value)
    }

    fun getShortValue(index: Int): Short {
        val value = get(index)
        try {
            return TypeUtils.castToShort(value)!!
        } catch (ignore: Exception) {
            return 0
        }
    }

    fun getInteger(index: Int): Int? {
        val value = get(index)
        return TypeUtils.castToInt(value)
    }

    fun getIntValue(index: Int): Int {
        val value = get(index)
        try {
            return TypeUtils.castToInt(value)!!
        } catch (ignore: Exception) {
            return 0
        }
    }

    fun getLong(index: Int): Long? {
        val value = get(index)
        return TypeUtils.castToLong(value)
    }

    fun getLongValue(index: Int): Long {
        val value = get(index)
        try {
            return TypeUtils.castToLong(value)!!
        } catch (ignore: Exception) {
            return 0L
        }
    }

    fun getFloat(index: Int): Float? {
        val value = get(index)
        return TypeUtils.castToFloat(value)
    }

    fun getFloatValue(index: Int): Float {
        val value = get(index)
        try {
            return TypeUtils.castToFloat(value)!!
        } catch (ignore: Exception) {
            return 0f
        }
    }

    fun getDouble(index: Int): Double? {
        val value = get(index)
        return TypeUtils.castToDouble(value)
    }

    fun getDoubleValue(index: Int): Double {
        val value = get(index)
        try {
            return TypeUtils.castToDouble(value)!!
        } catch (ignore: Exception) {
            return 0.0
        }
    }

    fun getBigDecimal(index: Int): BigDecimal? {
        val value = get(index)
        return TypeUtils.castToBigDecimal(value)
    }

    fun getBigInteger(index: Int): BigInteger? {
        val value = get(index)
        return TypeUtils.castToBigInteger(value)
    }

    fun getString(index: Int): String? {
        val value = get(index)
        return TypeUtils.castToString(value)
    }


    fun <T> getObject(index: Int, clazz: Class<T>): T? {
        val obj = list.get(index) ?: return null
        return JSON.parseObject(JSON.toJSONString(obj), clazz)
    }

    fun toJSONString(): String {
        return JSON.toJSONString(list)
    }

    override fun toString(): String {
        return toJSONString()
    }

    override val size: Int
        get() = list.size

    override fun isEmpty(): Boolean {
        return list.isEmpty()
    }

    override fun contains(element: Any?): Boolean {
        return list.contains(element)
    }

    override fun iterator(): MutableIterator<Any?> {
        return list.iterator()
    }

    fun toArray(): Array<Any?> {
        return list.toTypedArray()
    }

    override fun add(element: Any?): Boolean {
        return list.add(element)
    }

    override fun remove(element: Any?): Boolean {
        return list.remove(element)
    }

    override fun containsAll(elements: Collection<Any?>): Boolean {
        return list.containsAll(elements)
    }

    override fun addAll(elements: Collection<Any?>): Boolean {
        return list.addAll(elements)
    }

    override fun addAll(index: Int, elements: Collection<Any?>): Boolean {
        return list.addAll(index, elements)
    }

    override fun removeAll(elements: Collection<Any?>): Boolean {
        return list.removeAll(elements)
    }

    override fun retainAll(elements: Collection<Any?>): Boolean {
        return list.retainAll(elements)
    }

    override fun clear() {
        list.clear()
    }

    override fun equals(other: Any?): Boolean {
        return list == other
    }

    override fun hashCode(): Int {
        return list.hashCode()
    }

    override fun get(index: Int): Any? {
        return list[index]
    }

    override operator fun set(index: Int, element: Any?): Any? {
        return list.set(index, element)
    }

    override fun add(index: Int, element: Any?) {
        list.add(index, element)
    }

    override fun removeAt(index: Int): Any? {
        return list.removeAt(index)
    }

    override fun indexOf(element: Any?): Int {
        return list.indexOf(element)
    }

    override fun lastIndexOf(element: Any?): Int {
        return list.lastIndexOf(element)
    }

    override fun listIterator(): MutableListIterator<Any?> {
        return list.listIterator()
    }

    override fun listIterator(index: Int): MutableListIterator<Any?> {
        return list.listIterator(index)
    }

    override fun subList(fromIndex: Int, toIndex: Int): MutableList<Any?> {
        return list.subList(fromIndex, toIndex)
    }

    override fun replaceAll(operator: UnaryOperator<Any?>) {
        list.replaceAll(operator)
    }

    override fun sort(c: Comparator<in Any?>) {
        list.sortWith(c)
    }

    override fun spliterator(): Spliterator<Any?> {
        return list.spliterator()
    }
}
